### ISSUE TITLE
Installation issue: root fails to build with intel-oneapi-tbb

### ISSUE THREAD

**Issue by glennpj** [LLNL/spack#22616]
When intel-oneapi-tbb is used to provide tbb, the build of root fails.

### Steps to reproduce the issue

```console
$ spack spec root ^mesa@:20 | grep tbb
root@6.22.08%gcc@8.4.0~aqua+davix~emacs+examples~fftw~fits~fortran+gdml+gminimal~graphviz+gsl~http~ipo~jemalloc+math~memstat+minuit~mlp~mysql+opengl~postgres~pythia6~pythia8+python~qt4~r+roofit~root7+rpath~shadow~spectrum~sqlite~ssl~table+tbb+threads~tmva+unuran~vc+vdt~vmc+x+xml~xrootd build_type=RelWithDebInfo cxxstd=11 patches=22af3471f3fd87c0fe8917bf9c811c6d806de6c8b9867d30a1e3d383a1b929d7 arch=linux-centos7-broadwell
    ^intel-oneapi-tbb@2021.1.1%gcc@8.4.0 arch=linux-centos7-broadwell
```
```console
$ spack install root ^mesa@:20
...
1 error found in build log:
     96     -- Looking for LibXml2
     97     -- Found LibXml2: /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/libxml2-2.9.10-cgls2d3wuimv62r6pua43
            k443bxyhmv5/lib/libxml2.so (found version "2.9.10")
     98     -- Found FTGL: /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/ftgl-2.4.0-7wrzupbjlehdjrktbsaa6u3xtkls
            givc/include
     99     -- Checking for module 'davix>=0.6.4'
     100    --   Found davix, version 0.7.6
     101    -- Looking for TBB
  >> 102    CMake Error at /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/cmake-3.20.0-br6lwxtmpakaucaykabfl2hg46
            v5dfie/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
     103      Could NOT find TBB (missing: TBB_ROOT_DIR TBB_INCLUDE_DIR TBB_LIBRARY)
     104      (Required is at least version "2018")
     105    Call Stack (most recent call first):
     106      /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/cmake-3.20.0-br6lwxtmpakaucaykabfl2hg46v5dfie/share/
            cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
     107      cmake/modules/FindTBB.cmake:291 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
     108      cmake/modules/SearchInstalledSoftware.cmake:16 (_find_package)

```

### Information on your system

```console
$ spack debug report
* **Spack:** 0.16.1-1919-f59ac04
* **Python:** 3.6.8
* **Platform:** linux-centos7-broadwell
* **Concretizer:** original
```

### Additional information

root maintainers: @HadrienG2 @chissg @drbenmorgan @vvolkl 
intel-oneapi-tbb naintainers: @rscohn2 

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [ ] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by AlexVeprev** [comment]
Hi, I can provide support for oneTBB CMake part. Unfortunately, I'm not familiar with snack, so could you please clarify some moments? I see that the error comes from FindTBB.cmake that [is not distributed with the latest official CMake](https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules), so where this FindTBB.cmake comes from and where is it used?

For now I can assume that this FindTBB can't handle new oneTBB version (2021.1.1), because it is significantly changed in comparison with the previous one (TBB 2020), including some parts in layout.
So probably FindTBB should be updated to handle the new oneTBB properly or (even better) the approach should be changed from FindTBB to TBBConfig files (they are provided in _some_ oneTBB distributions). I can help with it.

**Comment by vvolkl** [comment]
Thanks @glennpj for the report and @AlexVeprev for the offer to help.

I'm not a TBB expert, but I guess `intel-oneapi-tbb` is  equivalent to `intel-tbb`? It is known that the 6.22 series of ROOT cannot handle newer `intel-tbb` version, which is why recently a conflict was added. See https://github.com/spack/spack/pull/22366 and the links in the comments. As far as I understand it, the ROOT developers are aware and are planning to fix it in newer versions.

Pinging also  @ChristianTackeGSI -- should we add an extra conflict for `intel-oneapi-tbb`?

**Comment by rscohn2** [comment]
I looked at #22366 , it is about use of functionality that is removed from TBB. This error looks like an issue with CMAKE finding TBB. The findtbb comes from here: https://github.com/root-project/root/blob/master/cmake/modules/FindTBB.cmake

They still might be blocked by #22366.

**Comment by AlexVeprev** [comment]
Thank you for the links!

> I guess intel-oneapi-tbb is equivalent to intel-tbb?

Actually, in this repo `intel-oneapi-tbb` is a new version of `intel-tbb`, it is the same product, but with major (incompatible) changes.

I see [intel-tbb/package.py](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/intel-tbb/package.py) is used for TBB 2020 and older versions and I see download/build/install instructions there. Also I see [intel-oneapi-tbb/package.py](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/intel-oneapi-tbb/package.py) for oneTBB 2021.1.1 (and probably future versions), it uses pre-built oneTBB from offline oneAPI installer. How is it unpacked and installed?

**Comment by rscohn2** [comment]
If you have spack installed. Then it is installed with:
```spack install intel-oneapi-tbb```
It will use the standalone installer and unpack into a non-default location. When building with tbb, it will source tbb/env/vars.sh. If setting some env variables will make the findtbb work better, we can add it to the tbb packages.

**Comment by AlexVeprev** [comment]
@rscohn2, thank you for the explanation! I've briefly checked FindTBB, having vars.sh sourced should be enough, but it is obviously not, so I need to reproduce the issue on my side to find the root cause.

**Comment by glennpj** [comment]
> It is known that the 6.22 series of ROOT cannot handle newer `intel-tbb` version, which is why recently a conflict was added.

Sorry, I missed that.

**Comment by AlexVeprev** [comment]
I've found that [FindTBB.cmake from root project](https://github.com/root-project/root/blob/master/cmake/modules/FindTBB.cmake) is not able to find TBB from `intel-oneapi-tbb` version because it has compiler subdir in the path to binaries while `intel-tbb` package doesn't have it. The FindTBB handles compiler subdirs on Windows, but doesn't on Linux: [FindTBB.cmake#L69](https://github.com/root-project/root/blob/master/cmake/modules/FindTBB.cmake#L69)

Binaries in `intel-tbb` package are located under: `<tbbroot>/lib`
Binaries in `intel-oneapi-tbb` packages are located under: `<tbbroot>/lib/intel64/gcc4.8` (FindTBB can handle `intel64` in the path, but can't handle `gcc4.8`.

BTW, root's FindTBB is not that modern approach to integrate TBB through CMake, now TBB provides more convenient and reliable integration ways. I could propose changes into root, but it'll require additional time to prepare and I'm not sure if it will help here right now.

**Comment by ChristianTackeGSI** [comment]
> > I guess intel-oneapi-tbb is equivalent to intel-tbb?
> 
> Actually, in this repo `intel-oneapi-tbb` is a new version of `intel-tbb`, it is the same product, but with major (incompatible) changes.

So, if I get that right
* `intel-tbb` is for anything up to `2020.*`.
* `intel-oneapi-tbb` is for 2021 and up.

So the conflicts in #22366 wont break ever.
But we should add a new one
```python
    conflicts('^intel-oneapi-tbb@2021.1:', when='@:6.22',
              msg='Please use an older intel-tbb/intel-oneapi-tbb version')
```

In theory, we could remove the conflict on `intel-tbb`, because it looks like there will never be a newer version. I'd prefer to keep it for documentation purposes.

Does that make sense to everybody?

**Comment by vvolkl** [comment]
@ChristianTackeGSI looks good to me!

@AlexVeprev I think that sounds great and I'm sure the ROOT developers would appreciate an updated `FindTBB.cmake`. They may have some constraints that are not obvious though - maybe start by opening an issue that mentions this problem and your plans, to give them a chance to comment?

**Comment by orlandini** [comment]
Hello,

I am joining the discussion because a colleague had problems with the newer TBB version in our own code. We provide our own `FindTBB.cmake`, written based on the modules that we have found online. It is far from perfect, but perhaps the search procedure could be interesting for you, so I decided to  [link it here](https://github.com/labmec/neopz/blob/RefactorGeom/cmake/Modules/FindTBB.cmake) in case it might be useful. I am sorry if it is not relevant.

**Comment by AlexVeprev** [comment]
>  maybe start by opening an issue that mentions this problem and your plans, to give them a chance to comment?

@vvolkl, yes, makes sense, thank you! I've found a similar issue in root https://github.com/root-project/root/issues/7063 and left a comment there.

**Comment by ChristianTackeGSI** [comment]
My suggested "fix" is now here: #22850
